### PR TITLE
Merge duplicates of source for empire.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This PR addresses https://github.com/freeorion/freeorion/issues/1242.

It also fixes a logic bug in that some universe objects were using invalid sources in their `ScriptingContexts`.

### CODE REVIEW

File: Empire/Empire.cpp [hunk 1]
```diff
@@ -1468,6 +1470,43 @@ int Empire::StockpileID(ResourceType res) const {
...
 
+int Empire::SourceID() const {
+    std::shared_ptr<const UniverseObject> good_source = Source();
+    return good_source ? good_source->ID() : INVALID_OBJECT_ID;
+}
+
+std::shared_ptr<const UniverseObject> Empire::Source() const {
+    // Make it safe to use with a null this pointer.
+    // This generates a compiler warning, but it enables the natural line of code of
+    // "GetEmpire(id)->Source()" without requiring an intervening null check
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: Just have a GetSourceForEmpire(int id) function that can return nullptr. Not having a check that GetEmpire(id) didn't return nullptr is weird / misleading.

File: Empire/Empire.cpp [hunk 2]
```diff
@@ -1480,8 +1519,10 @@ std::string Empire::Dump() const {
     return retval;
 }
 
-void Empire::SetCapitalID(int id)
-{ m_capital_id = id; }
+void Empire::SetCapitalID(int id) {
```
Comment (geoffthemedio) [inline Empire/Empire.cpp:1519]: probably shouldn't assume id is valid...

File: Empire/Empire.cpp [hunk 3]
```diff
@@ -1480,8 +1519,10 @@ std::string Empire::Dump() const {
...
+    m_capital_id = id;
+    m_source_id = id;
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: for purposes of setting the source id at least

File: Empire/Empire.h [hunk 4]
```diff
@@ -316,6 +317,14 @@ class FO_COMMON_API Empire {
     int                     CapitalID() const;                          ///< Returns the numeric ID of the empire's capital
     int                     StockpileID(ResourceType res) const;        ///< Returns the numeric ID of the empire's stockpile location for the resource of type \a res
 
+    /** Return an object id that is owned by the empire or
+        INVALID_OBJECT_ID. SourceID() is safe to call with a null this
+        pointer. */
+    int                     SourceID() const;
```
Comment (geoffthemedio) [inline Empire/Empire.h:321]: calling a member function on a null pointer is probably undefined behaviour... as above, no.

File: Empire/Empire.cpp (hunk 1 above)
Comment (adrianbroher) [reply Empire/Empire.cpp]: > +    // Make it safe to use with a null this pointer.

That's nonstandard undefined behaviour. Rejected.

**Review by adrianbroher — CHANGES_REQUESTED** [review]

File: Empire/Empire.cpp (hunk 1 above)
Comment (LGM-Doyle) [reply Empire/Empire.cpp]: I added EmpireManager::GetSource(id) and removed the no-standard behavior.

File: Empire/Empire.cpp (hunk 2 above)
Comment (LGM-Doyle) [reply Empire/Empire.cpp:1519]: I've added a check that the capital exists and is owned by the empire.

File: Empire/Empire.cpp (hunk 3 above)
Comment (LGM-Doyle) [reply Empire/Empire.cpp]: I've added a check that the source id is valid.  Source ids don't always need to exist in the code, it seems.

File: Empire/Empire.h (hunk 4 above)
Comment (LGM-Doyle) [reply Empire/Empire.h:321]: Fixed.

File: Empire/EmpireManager.cpp [hunk 5]
```diff
@@ -46,6 +46,11 @@ const Empire* EmpireManager::GetEmpire(int id) const {
     return it == m_empire_map.end() ? nullptr : it->second;
 }
 
+std::shared_ptr<const UniverseObject> EmpireManager::GetSource(int id) const {
+    auto it{m_empire_map.find(id)};
```
Comment (adrianbroher) [inline Empire/EmpireManager.cpp]: The OSX clang compiler spitted out error messages the last time I used a similar construct.

See 89f801cc3e

File: Empire/Empire.cpp [hunk 6]
```diff
@@ -1517,8 +1517,20 @@ std::string Empire::Dump() const {
...
-    m_capital_id = id;
-    m_source_id = id;
+    m_capital_id = INVALID_OBJECT_ID;
+    m_source_id = INVALID_OBJECT_ID;
+
+    if (id == INVALID_OBJECT_ID)
+        return;
+
+    // Verify that the capital exists and is owned by the empire
+    auto possible_capital {Objects().ExistingObject(id)};
```
Comment (adrianbroher) [inline Empire/Empire.cpp]: Same here.

> The OSX clang compiler spitted out error messages the last time I used a similar construct.
>
> See 89f801c

**Review by adrianbroher — APPROVED** [review]

File: Empire/EmpireManager.cpp (hunk 5 above)
Comment (LGM-Doyle) [reply Empire/EmpireManager.cpp]: New features, new quirks :smile:.  Fixed.

File: Empire/Empire.cpp (hunk 6 above)
Comment (LGM-Doyle) [reply Empire/Empire.cpp]: Fixed.

**Review by LGM-Doyle — COMMENTED** [review]
Thanks for the quick reviews.

**Comment by adrianbroher** [comment]
Looks good.

**Pull request merged**